Fix ArrayBuffer incorrectly reporting the upper bound in IndexOutOfBoundsException

mshibuya opened the pull request:
This is the fix scala/bug#12176.
~The solution suggested there seems to work~ , so I've just applied it and added partest for this.

Thanks in advance for reviewing!

mshibuya committed:
Fix ArrayBuffer incorrectly reporting the upper bound in IndexOutOfBoundsException

Fixes scala/bug#12176

mshibuya commented on src/library/scala/collection/mutable/ArrayBuffer.scala:
```diff
@@ -95,7 +95,7 @@ class ArrayBuffer[A] private (initialElements: Array[AnyRef], initialSize: Int)
   @inline private def checkWithinBounds(lo: Int, hi: Int) = {
     if (lo < 0) throw new IndexOutOfBoundsException(s"$lo is out of bounds (min 0, max ${size0-1})")
-    if (hi > size0) throw new IndexOutOfBoundsException(s"$hi is out of bounds (min 0, max ${size0 - 1})")
+    if (hi > size0) throw new IndexOutOfBoundsException(s"${hi - 1} is out of bounds (min 0, max ${size0 - 1})")
```
The solution suggested in scala/bug#12176 uses `$lo` here, but after a bit consideration it is wrong because it breaks with a situation where performing operations targeted for a range of indexes like this:
https://github.com/scala/scala/pull/9249/files#diff-421aa7b6b8733d6090f2148be711352af11cc34b2e6978c84051d2ecf0561bd6R27

Use of `$lo` produces incorrect message of `2 is out of bounds (min 0, max 2)`.

NthPortal commented:
the tests can probably just be junit tests rather than partests; you only really need to use partest if you need to check the compiler output/result

som-snytt commented:
The current tests do not check the message.

https://github.com/scala/scala/blob/2.13.x/test/junit/scala/collection/mutable/ArrayBufferTest.scala#L33

Random example checking exception message:

https://github.com/scala/scala/blob/2.13.x/test/junit/scala/reflect/internal/NamesTest.scala#L57

eed3si9n commented:
> the tests can probably just be junit tests rather than partests; you only really need to use partest if you need to check the compiler output/result

This PR was made during ScalaMatsuri PR. I should've thought about trying JUnit first, but I've probably mislead him since I only talked about partest.

som-snytt commented:
That is only natural, because we will miss partest during Scala 3. And to think how I hated partest at first! It's kind of like the "rom-com" of testing frameworks.

mshibuya commented:
Thanks for reviewing, I've moved the test into junit so please take a look!

diesalbla commented on test/junit/scala/collection/mutable/ArrayBufferTest.scala:
```diff
@@ -362,4 +362,22 @@ class ArrayBufferTest {
+    iiobe( newAB.update(-1, 'u'), "-1 is out of bounds (min 0, max 2)" )
+    iiobe( newAB.update(3, 'u'), "3 is out of bounds (min 0, max 2)" )
+    iiobe( newAB(3), "3 is out of bounds (min 0, max 2)" )
+    iiobe( newAB.remove(2, 2), "3 is out of bounds (min 0, max 2)" )
```
```suggestion
    iiobe( newAB.remove(3), "3 is out of bounds (min 0, max 2)" )
    iiobe( newAB.remove(2, 2), "3 is out of bounds (min 0, max 2)" )
```
Removing one and removing a block of several are separate operations.
https://www.scala-lang.org/api/current/scala/collection/mutable/ArrayBuffer.html#remove(index:Int):A
https://www.scala-lang.org/api/current/scala/collection/mutable/ArrayBuffer.html#remove(index:Int,count:Int):Unit

diesalbla replied:
https://www.scala-lang.org/api/current/scala/collection/mutable/ArrayBuffer.html#remove(index:Int):A

mshibuya committed:
Test remove(Int) and remove(Int, Int) separately

as suggested here:
https://github.com/scala/scala/pull/9249#discussion_r508456058

Co-authored-by: Diego E. Alonso Blas <diesalbla@gmail.com>

dwijnand reviewed (approved):
Thanks!
